projects/scullyio/ng-lib/src/lib/transfer-state/transfer-state.service.ts

```diff
@@ -0,0 +1,129 @@
+import { HttpClient } from '@angular/common/http';
+import { Inject, Injectable } from '@angular/core';
+import { DOCUMENT } from '@angular/common';
+import { NavigationStart, Router } from '@angular/router';
+import { isScullyGenerated, isScullyRunning } from '../utils/isScully';
+import { Observable, of, Subject } from 'rxjs';
+import { filter, map, switchMap, tap } from 'rxjs/operators';
+
+const SCULLY_SCRIPT_ID = `scully-transfer-state`;
+const SCULLY_STATE_START = `___SCULLY_STATE_START___`;
+const SCULLY_STATE_END = `___SCULLY_STATE_END___`;
+
+@Injectable({
+  providedIn: 'root',
+})
+export class TransferStateService {
+  private script: HTMLScriptElement;
+  private state: { [key: string]: any } = {};
+  private fetching: Subject<any>;
+
+  constructor(
+    @Inject(DOCUMENT) private document: Document,
+    private router: Router,
+    private http: HttpClient,
+  ) {
+    this.setupEnvForTransferState();
+    this.setupNavStartDataFetching();
+  }
+
+  private setupEnvForTransferState(): void {
+    if (isScullyRunning()) {
+      // In Scully puppeteer
+      this.script = this.document.createElement('script');
+      this.script.setAttribute('id', SCULLY_SCRIPT_ID);
+      this.script.setAttribute('type', `text/${SCULLY_SCRIPT_ID}`);
+      this.document.head.appendChild(this.script);
+    } else if (isScullyGenerated()) {
+      // On the client AFTER scully rendered it
+      this.script = this.document.getElementById(SCULLY_SCRIPT_ID) as HTMLScriptElement;
+      try {
+        this.state = JSON.parse(unescapeHtml(this.script.textContent));
+      } catch (e) {
+        this.state = {};
+      }
+    }
+  }
+
+  getState<T>(name: string): Observable<T> {
+    if (this.fetching) {
+      return this.fetching.pipe(map(() => this.state[name]));
+    } else {
+      return of(this.state[name]);
+    }
+  }
+
+  setState<T>(name: string, val: T): void {
+    this.state[name] = val;
+    if (isScullyRunning()) {
+      this.script.textContent = `${SCULLY_STATE_START}${escapeHtml(
```

Comment:
Inside a script tag, you don't really need to (un)escapeHtml, right?

Reply:
Since I escape it when I set it into the HTML, I may unescape it when I read it. 

This is what Universal does. Though I don't know why, I trust it solves something for them. That's why I did it too.

Reply:
@SanderElias you cool with leaving them in for now?